timeseries/src/main/scala/com/criteo/cuttle/timeseries/TimeSeriesScheduler.scala

```diff
@@ -420,41 +420,39 @@ case class TimeSeriesScheduler(logger: Logger) extends Scheduler[TimeSeries] wit
                                       runningExecutions: TMap[Execution[TimeSeries], Future[Completed]],
                                       xa: XA)(implicit user: Auth.User): IO[Either[String, Unit]] = {
 
+    logger.debug(s"Requesting a backfill of ${jobs.map(_.id)} between $start and $end")
 
-    val backfillErrors: List[String] = atomic { implicit txn =>
-      val errors = mutable.ArrayBuffer.empty[String]
-      val backfills = createBackfills(name, description, jobs, _state(), start, end, priority)
+    val currentJobStates = atomic { implicit txn => _state() }
 
-      val validBackfills: List[Backfill] = backfills.flatMap { newBackfill =>
-        validateBackfill(newBackfill, _state()).toList match {
-          case Nil =>  Some(newBackfill)
-          case (validationErrors) =>
-            errors += validationErrors.mkString("\n")
-            None
-        }
-      }
+    val backfillErrors = mutable.ArrayBuffer.empty[String]
+    val backfills = createBackfills(name, description, jobs, currentJobStates, start, end, priority)
 
-      if (errors.isEmpty) {
-        val executionsToCancel = listExecutionsToCancelGivenBackfills(validBackfills, _state(), runningExecutions.keys)
-        executionsToCancel.foreach { e =>
-          logger.info(s"Cancelling running execution ${e.id} for job ${e.job.id}")
-          e.cancel()
-        }
-        Future.sequence(executionsToCancel.map(runningExecutions.apply))
-          .andThen { case _ => validBackfills.foreach { newBackfill =>
-            updateBackfillState(newBackfill)
-            Database.createBackfill(newBackfill).transact(xa)
-          } }
+    val validBackfills: List[Backfill] = backfills.flatMap { newBackfill =>
+      validateBackfill(newBackfill, currentJobStates).toList match {
+        case Nil =>  Some(newBackfill)
+        case (validationErrors) =>
+          backfillErrors += validationErrors.mkString("\n")
+          None
       }
-      errors.toList
     }
 
-
-
     if (backfillErrors.nonEmpty)
       IO.pure(Left(backfillErrors.mkString("\n")))
-    else
+    else {
+      val (concurrentExecutionsToCancel, allExecutionCancelledPromise) = atomic { implicit txn =>
+        val executionsToCancel = listExecutionsToCancelGivenBackfills(validBackfills, currentJobStates, runningExecutions.keys)
+        (executionsToCancel, Future.sequence(executionsToCancel.map(runningExecutions.apply)))
```

Comment:
Here you try to cancel executions that you've constructed from state recuperated in a separate atomic block, it means that `executionsToCancel.map(runningExecutions.apply)` will throw if executionToCancel isn't runing anymore. And this will fail completely the Future.Sequence. Either treat this case with callbacks for each future either move this code in the same atomic as the validation. Personaly I don't think this is a good idea to separate the validation and cancellation code into two blocks

Reply:
The 'then' following the successful (or not as you pointed out) cancellation of executions will run no matter what (as we would like it to happen: don't block a backfill because we tried to cancel a still running execution which just finished).
I get your point indeed and will refactor the code to put everything into a same synchronization block.